concurrency/InternalComponentManager.java

```diff
@@ -0,0 +1,601 @@
+                routingTable.removeComponentRoute(componentJID);
+
+                // Ask the component to shutdown
+                component.shutdown();
+
+                if (!routingTable.hasComponentRoute(componentJID)) {
+                    // Remove the disco item from the server for the component that is being removed
+                    IQDiscoItemsHandler iqDiscoItemsHandler = XMPPServer.getInstance().getIQDiscoItemsHandler();
+                    if (iqDiscoItemsHandler != null) {
+                        iqDiscoItemsHandler.removeComponentItem(componentJID.toBareJID());
+                    }
+                    removeComponentInfo(componentJID);
+                    // Notify listeners that an existing component has been unregistered
+                    notifyComponentUnregistered(componentJID);
+                    // Alert other nodes of component removed event
+                    CacheFactory.doClusterTask(new NotifyComponentUnregistered(componentJID));
+                }
+                Log.debug("InternalComponentManager: Component unregistered for domain: " + subdomain);
+            }
+            else {
+                Log.debug("InternalComponentManager: Other components still tied to domain: " + subdomain);
+            }
+        }
+    }
+
+    void notifyComponentUnregistered(JID componentJID) {
+        for (ComponentEventListener listener : listeners) {
+            listener.componentUnregistered(componentJID);
+        }
+    }
+
+    void removeComponentInfo(JID componentJID) {
+        // Remove any info stored with the component being removed
+        componentInfo.remove(componentJID.getDomain());
+    }
+
+    @Override
+    public void sendPacket(Component component, Packet packet) {
+        if (packet != null && packet.getFrom() == null) {
+            throw new IllegalArgumentException("Packet with no FROM address was received from component.");
+        }
+        
+        PacketRouter router = XMPPServer.getInstance().getPacketRouter();
+        if (router != null) {
+            router.route(packet);
+        }
+    }
+
+    @Override
+    public IQ query(Component component, IQ packet, long timeout) throws ComponentException {
+        final LinkedBlockingQueue<IQ> answer = new LinkedBlockingQueue<>(8);
+        XMPPServer.getInstance().getIQRouter().addIQResultListener(packet.getID(), new IQResultListener() {
+            @Override
+            public void receivedAnswer(IQ packet) {
+                answer.offer(packet);
+            }
+
+            @Override
+            public void answerTimeout(String packetId) {
+                Log.warn("An answer to a previously sent IQ stanza was never received. Packet id: " + packetId);
+            }
+        });
+        sendPacket(component, packet);
+        IQ reply = null;
+        try {
+            reply = answer.poll(timeout, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)